A36-xds-for-servers.md

```diff
@@ -0,0 +1,385 @@
+based on `server.stop()`, but given C++ lacks this avenue the C API probably
+will not need this notification.
+
+
+### Java
+
+Create an `XdsServerBuilder` that extends `ServerBuilder` and delegates to a
+"real" builder. The server returned from `builder.build()` would be an xDS-aware
+server delegating to a "real" server instance. The `XdsServerBuilder` would
+install an `AtomicReference<ServerInterceptor>`-backed interceptor in the
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once; `build()` should throw if called a
+second time. To allow the `XdsServerBuilder` to create addition servers if
+necessary, mutation of the builder after `build()` should also throw.
+
+The xDS-aware server’s `start()` will create the `XdsClient` with the passed
+port and wait for initial configuration before delegating to the real `start()`.
+If the xDS bootstrap is missing it will throw an IOException within `start()`.
+The `XdsClient` will be shut down when the server is terminated (generally
+noticed via `shutdownNow()`/`awaitTermination()`). When new or updated
+configuration is received, it will create a new intercepter and update the
+interceptor reference.
+
+XdsServerBuilder will have an
+`xdsServingStatusListener(XdsServingStatusListener)` method.
+`XdsServingStatusListener` will be defined as:
+
+```java
+package io.grpc.xds;
+
+public interface XdsServingStatusListener {
+  void onServing();
+  void onNotServing(Throwable t);
+}
+```
+
+It is an interface instead of an abstract class as additional methods are not
+expected to be added before Java 8 language features are permitted in the code
+base. If this proves incorrect, an additional interface can be added.
+
+If not specified, a default implementation of `XdsServingStatusListener` will be
+used. It will log the exception and log calls to `onServing()` following a call
+to `onNotServing()` at WARNING level.
+
+In order to allow transport-specific configuration, the `XdsServerBuilder` will
+have a `@ExperimentalApi ServerBuilder transportBuilder()` method whose return
+value can be cast to `NettyServerBuilder` for experimental API configuration.
+
+`NettyServerBuilder` will add an internal method `eagAttributes(Attributes)`. It
+will plumb those attributes to `GrpcHttp2ConnectionHandler.getEagAttributes()`
+as implemented by `NettyServerHandler`, which currently just returns
+`Attributes.EMPTY`. `XdsServerBuilder` will then specify `eagAttributes` to
+inject credential information for the `XdsServerCredential`, similar to
+client-side, although with the added step of needing to process the
+FilterChainMatch for the specific connection.
+
+
+### Go
+
+Create an `xds.GRPCServer` struct that would internally contain an unexported
+`grpc.Server`. It would inject its own `StreamServerInterceptor` and
```

Comment:
@easwars, if a user has their own interceptor, how will this interceptor call their interceptor?